File: sdks/typescript/packages/core/src/engine-downloader.ts
import * as https from "node:https";
import * as http from "node:http";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import * as crypto from "node:crypto";
import { ENGINE_VERSION } from "./version.js";

const RELEASE_BASE = `https://github.com/attest-framework/attest/releases/download/v${ENGINE_VERSION}`;
const CHECKSUMS_URL = `${RELEASE_BASE}/checksums-sha256.txt`;
const MAX_REDIRECTS = 5;

export function attestBinDir(): string {
  const dir = path.join(os.homedir(), ".attest", "bin");
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

export function platformKey(): string {
  const platform = os.platform();
  const arch = os.arch();

  const mapping: Record<string, Record<string, string>> = {
    darwin: { arm64: "darwin-arm64", x64: "darwin-amd64" },
    linux: { arm64: "linux-arm64", x64: "linux-amd64" },
    win32: { x64: "windows-amd64", arm64: "windows-arm64" }
  };

  const osMap = mapping[platform];
  if (!osMap) {
    throw new Error(
      `Unsupported platform: ${platform}. Supported: darwin, linux, win32.`
    );
  }

  const key = osMap[arch];
  if (!key) {
    throw new Error(
      `Unsupported architecture '${arch}' on platform '${platform}'. ` +
      `Supported architectures: ${Object.keys(osMap).join(", ")}.`
    );
  }

  return key;
}

export function binaryFilename(): string {
  return os.platform() === "win32" ? "attest-engine.exe" : "attest-engine";
}

export function cachedEnginePath(): string | null {
  const dir = path.join(os.homedir(), ".attest", "bin");
  const binPath = path.join(dir, binaryFilename());
  const versionFile = path.join(dir, ".engine-version");

  if (!fs.existsSync(binPath)) {
    return null;
  }

  if (!fs.existsSync(versionFile)) {
    return null;
  }

  const cachedVersion = fs.readFileSync(versionFile, "utf-8").trim();
  if (cachedVersion !== ENGINE_VERSION) {
    return null;
  }

  return binPath;
}

export function parseChecksums(text: string): Map<string, string> {
  const checksums = new Map<string, string>();
  for (const line of text.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    // Format: "{hash}  {filename}" (two spaces between hash and filename)
    const match = trimmed.match(/^([a-f0-9]{64})\s+(.+)$/);
    if (match) {
      checksums.set(match[2], match[1]);
    }
  }
  return checksums;
}

function httpsGet(url: string, redirectCount = 0): Promise<Buffer> {
  if (redirectCount > MAX_REDIRECTS) {
    return Promise.reject(
      new Error(`Too many redirects (>${MAX_REDIRECTS}) following ${url}`)
    );
  }

  return new Promise<Buffer>((resolve, reject) => {
    const client = url.startsWith("https:") ? https : http;
    client.get(url, (res) => {
      const status = res.statusCode ?? 0;

      if (status === 301 || status === 302) {
        const location = res.headers.location;
        if (!location) {
          reject(new Error(`Redirect ${status} without Location header from ${url}`));
          return;
        }
        res.resume();
        resolve(httpsGet(location, redirectCount + 1));
        return;
      }

      if (status < 200 || status >= 300) {
        res.resume();
        reject(new Error(`HTTP ${status} fetching ${url}`));
        return;
      }

      const chunks: Buffer[] = [];
      res.on("data", (chunk: Buffer) => chunks.push(chunk));
      res.on("end", () => resolve(Buffer.concat(chunks)));
      res.on("error", reject);
    }).on("error", reject);
  });
}

export async function downloadEngine(): Promise<string> {
  const key = platformKey();
  const filename = `attest-engine-${key}`;
  const binaryUrl = `${RELEASE_BASE}/${filename}`;

  process.stderr.write(`[attest] Downloading engine v${ENGINE_VERSION} for ${key}...\n`);

  // Fetch checksums
  process.stderr.write(`[attest] Fetching checksums...\n`);
  const checksumsText = (await httpsGet(CHECKSUMS_URL)).toString("utf-8");
  const checksums = parseChecksums(checksumsText);
  const expectedHash = checksums.get(filename);
  if (!expectedHash) {
    throw new Error(
      `No checksum found for '${filename}' in checksums-sha256.txt. ` +
      `Available files: ${[...checksums.keys()].join(", ")}`
    );
  }

  // Download binary
  process.stderr.write(`[attest] Downloading ${binaryUrl}...\n`);
  const data = await httpsGet(binaryUrl);

  // Verify SHA256
  const actualHash = crypto.createHash("sha256").update(data).digest("hex");
  if (actualHash !== expectedHash) {
    throw new Error(
      `SHA256 mismatch for ${filename}. ` +
      `Expected: ${expectedHash}, got: ${actualHash}. ` +
      `The download may be corrupted. Delete ~/.attest/bin/ and retry.`
    );
  }
  process.stderr.write(`[attest] Checksum verified.\n`);

  // Atomic write: temp file + rename
  const dir = attestBinDir();
  const destFilename = binaryFilename();
  const destPath = path.join(dir, destFilename);
  const tmpPath = path.join(dir, `.${destFilename}.tmp.${process.pid}`);

  fs.writeFileSync(tmpPath, data);
  fs.chmodSync(tmpPath, 0o755);
  fs.renameSync(tmpPath, destPath);

  // Write version marker
  fs.writeFileSync(path.join(dir, ".engine-version"), ENGINE_VERSION);

  process.stderr.write(`[attest] Engine installed at ${destPath}\n`);
  return destPath;
}

File: sdks/python/src/attest/engine_downloader.py
"""Auto-download attest-engine binary from GitHub Releases."""

from __future__ import annotations

import hashlib
import os
import platform
import stat
import sys
import tempfile
import urllib.error
import urllib.request
from pathlib import Path

from attest import ENGINE_VERSION

_GITHUB_RELEASE_BASE = (
    "https://github.com/attest-framework/attest/releases/download"
)

_PLATFORM_MAP: dict[str, str] = {
    "Darwin-arm64": "darwin-arm64",
    "Darwin-x86_64": "darwin-amd64",
    "Linux-aarch64": "linux-arm64",
    "Linux-x86_64": "linux-amd64",
    "Windows-AMD64": "windows-amd64",
    "Windows-ARM64": "windows-arm64",
}


def _attest_bin_dir() -> Path:
    """Return ``~/.attest/bin/``, creating it if absent."""
    bin_dir = Path.home() / ".attest" / "bin"
    bin_dir.mkdir(parents=True, exist_ok=True)
    return bin_dir


def _platform_key() -> str:
    """Map current OS/arch to the release asset naming convention."""
    system = platform.system()
    machine = platform.machine()
    key = f"{system}-{machine}"
    mapped = _PLATFORM_MAP.get(key)
    if mapped is None:
        raise RuntimeError(
            f"Unsupported platform: {key}. "
            f"Supported: {', '.join(sorted(_PLATFORM_MAP.values()))}"
        )
    return mapped


def _binary_filename() -> str:
    """Return the engine binary filename for the current OS."""
    if platform.system() == "Windows":
        return "attest-engine.exe"
    return "attest-engine"


def _version_file() -> Path:
    """Return the path to the cached engine version marker."""
    return _attest_bin_dir() / ".engine-version"


def cached_engine_path() -> Path | None:
    """Return the cached binary path if it exists and matches ENGINE_VERSION."""
    bin_path = _attest_bin_dir() / _binary_filename()
    if not bin_path.is_file():
        return None

    ver_file = _version_file()
    if not ver_file.is_file():
        return None

    cached_version = ver_file.read_text().strip()
    if cached_version != ENGINE_VERSION:
        return None

    return bin_path


def _parse_checksums(text: str) -> dict[str, str]:
    """Parse a checksums file with ``{hash}  {filename}`` lines."""
    checksums: dict[str, str] = {}
    for line in text.splitlines():
        line = line.strip()
        if not line:
            continue
        # Format: sha256hash  filename (two spaces)
        parts = line.split(None, 1)
        if len(parts) == 2:
            checksums[parts[1].strip()] = parts[0]
    return checksums


def _url_read(url: str) -> bytes:
    """Fetch a URL, following redirects. Stdlib-only."""
    request = urllib.request.Request(url, headers={"User-Agent": "attest-sdk"})
    with urllib.request.urlopen(request, timeout=120) as resp:
        return resp.read()  # type: ignore[no-any-return]


def download_engine() -> Path:
    """Download the engine binary from GitHub Releases with SHA256 verification.

    Returns the path to the downloaded binary.

    Raises:
        RuntimeError: On download failure, checksum mismatch, or unsupported platform.
    """
    plat = _platform_key()
    ver = ENGINE_VERSION
    bin_name = _binary_filename()
    asset_name = f"attest-engine-{plat}"
    if platform.system() == "Windows":
        asset_name += ".exe"

    binary_url = f"{_GITHUB_RELEASE_BASE}/v{ver}/{asset_name}"
    checksums_url = f"{_GITHUB_RELEASE_BASE}/v{ver}/checksums-sha256.txt"

    sys.stderr.write(f"attest: downloading engine v{ver} for {plat}...\n")

    # Fetch checksums first
    try:
        checksums_text = _url_read(checksums_url).decode("utf-8")
    except (urllib.error.URLError, OSError) as exc:
        raise RuntimeError(
            f"Failed to download checksums from {checksums_url}: {exc}\n"
            "Verify the release exists at "
            f"https://github.com/attest-framework/attest/releases/tag/v{ver}"
        ) from exc

    checksums = _parse_checksums(checksums_text)
    expected_hash = checksums.get(asset_name)
    if expected_hash is None:
        raise RuntimeError(
            f"No checksum found for '{asset_name}' in checksums-sha256.txt. "
            f"Available assets: {', '.join(sorted(checksums.keys()))}"
        )

    # Download binary
    try:
        binary_data = _url_read(binary_url)
    except (urllib.error.URLError, OSError) as exc:
        raise RuntimeError(
            f"Failed to download engine from {binary_url}: {exc}"
        ) from exc

    # Verify SHA256
    actual_hash = hashlib.sha256(binary_data).hexdigest()
    if actual_hash != expected_hash:
        raise RuntimeError(
            f"SHA256 mismatch for {asset_name}:\n"
            f"  expected: {expected_hash}\n"
            f"  actual:   {actual_hash}\n"
            "The download may be corrupted. Retry or download manually."
        )

    # Atomic write: temp file + rename
    bin_dir = _attest_bin_dir()
    target = bin_dir / bin_name

    fd, tmp_path = tempfile.mkstemp(dir=bin_dir, prefix=".attest-engine-tmp-")
    try:
        with os.fdopen(fd, "wb") as f:
            f.write(binary_data)
        os.chmod(tmp_path, stat.S_IRWXU | stat.S_IRGRP | stat.S_IXGRP | stat.S_IROTH | stat.S_IXOTH)
        os.replace(tmp_path, target)
    except BaseException:
        Path(tmp_path).unlink(missing_ok=True)
        raise

    # Write version marker atomically
    ver_file = _version_file()
    fd_v, tmp_ver = tempfile.mkstemp(dir=bin_dir, prefix=".engine-version-tmp-")
    try:
        with os.fdopen(fd_v, "w") as f:
            f.write(ver)
        os.replace(tmp_ver, ver_file)
    except BaseException:
        Path(tmp_ver).unlink(missing_ok=True)
        raise

    sys.stderr.write(f"attest: engine v{ver} installed to {target}\n")
    return target
